File: javascript/src/entities/car.js
// Car entity with modern JavaScript features
import { TRACK_LENGTH, NUM_LANES, CAR_SYMBOLS } from '../config.js';

/**
 * Car class representing a racing vehicle
 * Demonstrates: ES6 classes, private fields, getters/setters, async/await
 */
export class Car {
  // Private fields (ES2022+)
  #speed;
  #controller;

  /**
   * @param {number} id - Unique car identifier
   * @param {number} lane - Starting lane (0-3)
   * @param {Object} controller - Controller instance for decision making
   */
  constructor(id, lane, controller) {
    this.id = id;
    this.lane = lane;
    this.position = 0;
    this.#speed = Math.random() * 1.5 + 0.5; // Random speed between 0.5 and 2.0
    this.symbol = CAR_SYMBOLS[id % CAR_SYMBOLS.length];
    this.status = 'active'; // 'active', 'crashed', 'finished'
    this.#controller = controller;
  }

  /**
   * Getter for speed (private field access)
   */
  get speed() {
    return this.#speed;
  }

  /**
   * Move the car down the track
   */
  moveDown() {
    if (this.status !== 'active') return;

    this.position += this.#speed;

    // Check if finished
    if (this.position >= TRACK_LENGTH) {
      this.position = TRACK_LENGTH;
      this.status = 'finished';
    }
  }

  /**
   * Switch to a different lane
   * @param {number} direction - Direction to move (-1 for left, 1 for right)
   */
  switchLane(direction) {
    if (this.status !== 'active') return;

    const newLane = this.lane + direction;

    // Boundary checking
    if (newLane >= 0 && newLane < NUM_LANES) {
      this.lane = newLane;
    }
  }

  /**
   * Check collision with obstacles
   * @param {Array} obstacles - Array of obstacle objects
   * @returns {boolean} - True if collision detected
   */
  checkCollision(obstacles) {
    if (this.status !== 'active') return false;

    // Check if any obstacle is at the same position and lane
    return obstacles.some(obstacle =>
      obstacle.lane === this.lane &&
      Math.abs(obstacle.position - this.position) < 1
    );
  }

  /**
   * Handle collision event
   */
  crash() {
    this.status = 'crashed';
    this.#speed *= 0.5; // Reduce speed after crash
  }

  /**
   * Get current state for controller decision making
   * @returns {Object} - Car state snapshot
   */
  getState() {
    return {
      id: this.id,
      lane: this.lane,
      position: this.position,
      speed: this.#speed,
      status: this.status
    };
  }

  /**
   * Async update method - gets decision from controller and executes it
   * Demonstrates async/await pattern
   */
  async updateAsync() {
    if (this.status !== 'active') return;

    // Get decision from controller (strategy pattern)
    const action = await this.#controller.decideAction(this.getState());

    // Execute the decision
    if (action.type === 'lane_change') {
      this.switchLane(action.direction);
    }

    // Move down the track
    this.moveDown();
  }

  /**
   * Serialize car state to plain object
   * @returns {Object} - Serialized car data
   */
  toObject() {
    return {
      id: this.id,
      lane: this.lane,
      position: this.position,
      speed: this.#speed,
      symbol: this.symbol,
      status: this.status
    };
  }
}

File: python/src/entities/car.py
"""Car entity with modern Python features"""

import random
from typing import Any, Protocol
from config import TRACK_LENGTH, NUM_LANES, CAR_SYMBOLS


class Controller(Protocol):
    """Protocol defining the controller interface"""
    async def decide_action(self, car_state: dict[str, Any]) -> dict[str, Any]:
        """Decide what action to take based on car state"""
        ...


class Car:
    """
    Car class representing a racing vehicle
    Demonstrates: Type hints, properties, private attributes, async/await
    """

    def __init__(self, id: int, lane: int, controller: Controller) -> None:
        """
        Initialize a car instance

        Args:
            id: Unique car identifier
            lane: Starting lane (0-3)
            controller: Controller instance for decision making
        """
        self.id: int = id
        self.lane: int = lane
        self.position: float = 0.0
        self._speed: float = random.random() * 1.5 + 0.5  # Random speed between 0.5 and 2.0
        self.symbol: str = CAR_SYMBOLS[id % len(CAR_SYMBOLS)]
        self.status: str = 'active'  # 'active', 'crashed', 'finished' TODO: does python have a pythonic enum
        self._controller: Controller = controller

    @property
    def speed(self) -> float:
        """Get car speed (property decorator for encapsulation)"""
        return self._speed

    def move_down(self) -> None:
        """Move the car down the track"""
        if self.status != 'active':
            return

        self.position += self._speed

        # Check if finished
        if self.position >= TRACK_LENGTH:
            self.position = TRACK_LENGTH
            self.status = 'finished'

    def switch_lane(self, direction: int) -> None:
        """
        Switch to a different lane

        Args:
            direction: Direction to move (-1 for left, 1 for right)
        """
        if self.status != 'active':
            return

        new_lane = self.lane + direction

        # Boundary checking
        if 0 <= new_lane < NUM_LANES:
            self.lane = new_lane

    def check_collision(self, obstacles: list[Any]) -> bool:
        """
        Check collision with obstacles

        Args:
            obstacles: List of obstacle objects

        Returns:
            True if collision detected
        """
        if self.status != 'active':
            return False

        # Check if any obstacle is at the same position and lane
        return any(
            obstacle.lane == self.lane and abs(obstacle.position - self.position) < 1
            for obstacle in obstacles
        )

    def crash(self) -> None:
        """Handle collision event"""
        self.status = 'crashed'
        self._speed *= 0.5  # Reduce speed after crash

    def get_state(self) -> dict[str, Any]:
        """
        Get current state for controller decision making

        Returns:
            Car state snapshot
        """
        return {
            'id': self.id,
            'lane': self.lane,
            'position': self.position,
            'speed': self._speed,
            'status': self.status
        }

    async def update_async(self) -> None:
        """
        Async update method - gets decision from controller and executes it
        Demonstrates async/await pattern
        """
        if self.status != 'active':
            return

        # Get decision from controller (strategy pattern)
        action = await self._controller.decide_action(self.get_state())

        # Execute the decision
        if action.get('type') == 'lane_change':
            self.switch_lane(action['direction'])

        # Move down the track
        self.move_down()

    def to_dict(self) -> dict[str, Any]:
        """
        Serialize car state to dictionary

        Returns:
            Serialized car data
        """
        return {
            'id': self.id,
            'lane': self.lane,
            'position': self.position,
            'speed': self._speed,
            'symbol': self.symbol,
            'status': self.status
        }
